### PR TITLE
WIP: add catboost 

### PR DESCRIPTION
**Pull request by eugen-ajechiloae-clearml**

### CODE REVIEW

File: clearml/binding/frameworks/catboost_bind.py [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+
+    @staticmethod
+    def _generate_training_callback_class():
+        class ClearMLCallback:
+            def __init__(self, task):
+                self._logger = task.get_logger()
+
+            def after_iteration(self, info):
+                info = vars(info)
+                iteration = info['iteration']
```
Comment (JDennisJ) [inline clearml/binding/frameworks/catboost_bind.py:105]: `info.get` is safer in this case

**Review by JDennisJ — CHANGES_REQUESTED** [review]
Looks good :) Update me if you need any help with the comments

Also please run black for formatting (max-line-length=120) :)

File: clearml/binding/frameworks/catboost_bind.py [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+            # noinspection PyBroadException
+            try:
+                model.trains_in_model = empty.trains_in_model
+            except Exception:
+                pass
+        return model
+
+    @staticmethod
+    def _fit(original_fn, obj, *args, **kwargs):
+        print('got in _callback_after_iteration')
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply clearml/binding/frameworks/catboost_bind.py]: removed

File: clearml/binding/frameworks/catboost_bind.py [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+        try:
+            model_name = Path(filename).stem
+        except Exception:
+            model_name = None
+        WeightsFileHandler.create_output_model(obj, filename, Framework.catboost, PatchCatBoostModelIO.__main_task,
+                                               singlefile=True, model_name=model_name)
+        return ret
+
+    @staticmethod
+    def _load(original_fn, obj, f, *args, **kwargs):
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply clearml/binding/frameworks/catboost_bind.py]: function now has the same signature as in PatchBaseModelIO

File: clearml/binding/frameworks/catboost_bind.py [hunk 4]
```diff
@@ -0,0 +1,111 @@
+import sys
+
+import six
+from pathlib2 import Path
+
+from ..frameworks.base_bind import PatchBaseModelIO
+from ..frameworks import _patched_call, WeightsFileHandler, _Empty
+from ..import_bind import PostImportHookPatching
+from ...config import running_remotely
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply clearml/binding/frameworks/catboost_bind.py]: removed

File: clearml/binding/frameworks/catboost_bind.py [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+            CatBoost.save_model = _patched_call(CatBoost.save_model, PatchCatBoostModelIO._save)
+            CatBoost.load_model = _patched_call(CatBoost.load_model, PatchCatBoostModelIO._load)
+            PatchCatBoostModelIO.__callback_cls = PatchCatBoostModelIO._generate_training_callback_class()
+            CatBoost.fit = _patched_call(CatBoost.fit, PatchCatBoostModelIO._fit)
+            CatBoostClassifier.fit = _patched_call(CatBoostClassifier.fit, PatchCatBoostModelIO._fit)
+            CatBoostRegressor.fit = _patched_call(CatBoostRegressor.fit, PatchCatBoostModelIO._fit)
+            CatBoostRanker.fit = _patched_call(CatBoostRegressor.fit, PatchCatBoostModelIO._fit)
+        except ImportError:
+            pass
+        except Exception:
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply clearml/binding/frameworks/catboost_bind.py:43]: added

File: examples/frameworks/catboost/catboost_example.py [hunk 6]
```diff
@@ -0,0 +1,53 @@
+import numpy as np
+from sklearn.linear_model import Ridge
```
[the remark this answers is not in the document]
Comment (eugen-ajechiloae-clearml) [reply examples/frameworks/catboost/catboost_example.py]: removed

File: clearml/binding/frameworks/catboost_bind.py (hunk 1 above)
Comment (eugen-ajechiloae-clearml) [reply clearml/binding/frameworks/catboost_bind.py]: using info.get now

**Comment by erezalg** [comment]
In the example's requirement.txt there are a lot of external lib imports, I suggest removing:
click, boto3, IPython, ipywidgets 
unless they are really necessary. 
BTW, if they are necessary, consider a simpler example code (instead of click use argparse, not use boto3 and not sure why would you need IPython for an example code)

**Pull request merged**